katdal/chunkstore.py

```diff
@@ -437,6 +441,58 @@ def put_dask_array(self, array_name, array, offset=()):
         out_chunks = tuple(len(c) * (1,) for c in array.chunks)
         return da.Array(dask_graph, out_name, out_chunks, np.object)
 
+    def list_chunk_ids(self, array_name):
+        """List all chunk ID strings associated with given array in chunk store.
+
+        Parameters
+        ----------
+        array_name : string
+            Identifier of array in chunk store
+
+        Returns
+        -------
+        chunk_ids : list of string
+            List of chunk identifier strings (e.g. '00012_01024_00000')
+        """
+        raise NotImplementedError
+
+    def has_array(self, array_name, chunks, offset=()):
+        """Check if array is in the store.
+
+        This is an optional optimised version of :meth:`has_dask_array`.
+
+        Parameters
+        ----------
+        array_name : string
+            Identifier of array in chunk store
+        chunks : tuple of tuples of ints
+            Chunk specification
+        offset : tuple of int, optional
+            Offset to add to each dimension when addressing chunks in store
+
+        Returns
+        -------
+        success : :class:`numpy.ndarray` object
+            Array of bools indicating presence of each chunk
+
+        Raises
+        ------
+        NotImplementedError
+            If the underlying store does not have an efficient implementation
+        """
+        # Obtain ID strings of all chunks in store associated with array_name
+        # This might not be implemented by underlying store
+        store_ids = set(self.list_chunk_ids(array_name))
+        # Turn chunks + offset into list of expected chunk ID strings
+        slices = da.core.slices_from_chunks(chunks)
+        if offset:
+            offset_slices = tuple(slice(start, None) for start in offset)
+            slices = [da.core.fuse_slice(offset_slices, s) for s in slices]
```

Comment:
How much does this add to runtime? fuse_slice seems like a rather large hammer for this, but if it's not slowing things down then it's fine.

Reply:
I don't know. The offsets are not used much - mostly in `vis_writer` where each heap is chunked up separately and written into the larger array. Oh wait, you ditched `put_dask_array` :-)

I see how you did this in `vis_writer` - I'll pilfer that then...